extensions/amp-story/1.0/amp-story.js

```diff
@@ -1281,17 +1305,48 @@ export class AmpStory extends AMP.BaseElement {
       return Promise.resolve();
     }
 
+    const subscriptionsState = this.storeService_.get(
+      StateProperty.SUBSCRIPTIONS_STATE
+    );
     if (
-      isExperimentOn(this.win, 'amp-story-paywall-exp') &&
-      targetPage.isPaywallProtected()
+      isExperimentOn(this.win, 'amp-story-subscriptions') &&
+      this.isPaywallStory_()
     ) {
-      if (this.storeService_.get(StateProperty.SUBSCRIPTIONS_DIALOG_STATE)) {
-        // Subscription dialog is already triggered.
-        return Promise.resolve();
+      if (
+        pageIndex >= PAYWALL_PAGE_INDEX &&
+        subscriptionsState !== SubscriptionsState.GRANTED
+      ) {
+        // Hit a blocked page.
+        if (subscriptionsState === SubscriptionsState.UNKNOWN) {
+          // Block while waiting for entitlements.
+          return this.subscriptionsStatePromise_.promise.then(() => {
+            return this.switchTo_(targetPageId, direction);
+          });
+        } else {
+          // Show the paywall if the access is not granted.
+          if (pageIndex === PAYWALL_PAGE_INDEX) {
+            this.paywallTimeout_ = setTimeout(() => {
+              this.storeService_.dispatch(
+                Action.TOGGLE_SUBSCRIPTIONS_DIALOG_UI_STATE,
+                true
+              );
+            }, PAYWALL_DELAY_DURATION);
+          } else {
+            this.storeService_.dispatch(
+              Action.TOGGLE_SUBSCRIPTIONS_DIALOG_UI_STATE,
+              true
+            );
+            return Promise.resolve();
+          }
+        }
+      } else {
+        // Reset paywall UI if visting a non-blocked page.
+        this.paywallTimeout_ && clearTimeout(this.paywallTimeout_);
+        this.storeService_.dispatch(
+          Action.TOGGLE_SUBSCRIPTIONS_DIALOG_UI_STATE,
```

Comment:
How are we hiding the UI if a user is `BLOCKED` then becomes `GRANTED`? I think this dispatch needs to be outside the top level `if`. Also please add test for this

Reply:
My mistake I think I was looking at wrong level of indentation here.

Reply:
This logic of from blocked to granted -> hide UI was in amp-story-subscription, but now you mention it I think that logic would be better here since we already registered a handler for subscription state. Updated the test(actually already have this test case as 'should be able to navigate to all locked pages once status becomes granted') and moved the logic.

For dispatch here, this is for like navigate back to the previous story or blocked page with granted status to hide the UI. So this else block basically cover all non-blocked switchTo cases for a paywall story.